### PR TITLE
feat(expect): expose expect timeout

### LINKED ISSUE #30583: [Feature]: Accessing the expect timeout configuration from custom matcher
### 🚀 Feature Request

Hi Team.
I'm making a custom matcher and would like it make it `Auto-retrying assertions`

I tried to access expect timeout using `test.info()` but I didn't see `expect` property.

Is there any way to access it in a custom matcher?

image

### Example

_No response_

### Motivation

User can create `Auto-retrying assertions` aligned with the timeout setting in configuration

### PR DESCRIPTION
**Pull request by yury-s**
Fixes https://github.com/microsoft/playwright/issues/30583

### CODE REVIEW

File: packages/playwright/src/matchers/expect.ts [hunk 1]
```diff
@@ -129,7 +129,14 @@ function createExpect(info: ExpectMetaInfo) {
 
       if (property === 'extend') {
         return (matchers: any) => {
-          expectLibrary.extend(matchers);
+          const wrappedMatchers: any = {};
+          Object.entries(matchers).forEach(([name, matcher]) => {
+            wrappedMatchers[name] = function (...args: any[]) {
+              this.timeout = currentExpectTimeout({});
```
Comment (dgozman) [inline packages/playwright/src/matchers/expect.ts:135]: As discussed offline, let's create a new `this` object here, and also update our matchers to use `this.timeout`.

File: packages/playwright/src/matchers/toMatchText.ts [hunk 2]
```diff
@@ -49,17 +49,17 @@ export async function toMatchText(
     !(expected && typeof expected.test === 'function')
   ) {
     throw new Error(
-        this.utils.matcherErrorMessage(
+        matcherErrorMessage(
```
Comment (dgozman) [inline packages/playwright/src/matchers/toMatchText.ts:52]: Looking at the `matcherErrorMessage` implementation, perhaps just inline it? We can also inline COLORs in a separate PR.

File: utils/generate_types/overrides-test.d.ts [hunk 3]
```diff
@@ -361,6 +361,8 @@ export type ExpectMatcherState = {
   isNot: boolean;
   promise: 'rejects' | 'resolves' | '';
   utils: ExpectMatcherUtils;
+  expand: boolean;
```
Comment (dgozman) [inline utils/generate_types/overrides-test.d.ts:364]: - What is this property?
- Let's manually add JSDocs to properties here? At least for `isNot` and `timeout`.

File: tests/playwright-test/expect.spec.ts [hunk 4]
```diff
@@ -1000,3 +1000,42 @@ test('should respect timeout from configured expect when used outside of the tes
...
+    'playwright.config.ts': `
+      export default {
+        expect: { timeout: 1100 }
+      };
+    `,
+    'a.test.ts': `
+      import type { ExpectMatcherState, MatcherReturnType } from '@playwright/test';
+      import { test, expect as base } from '@playwright/test';
+
+      const expect = base.extend<{assertTimeout: (this: ExpectMatcherState, page: any, value: number) => MatcherReturnType}>({
```
Comment (dgozman) [inline tests/playwright-test/expect.spec.ts:1015]: Typescript should infer things:

```suggestion
      const expect = base.extend({
```

File: utils/generate_types/overrides-test.d.ts (hunk 3 above)
Comment (yury-s) [reply utils/generate_types/overrides-test.d.ts:364]: > * What is this property?

Removed.

> * Let's manually add JSDocs to properties here? At least for `isNot` and `timeout`.

Done.

File: utils/generate_types/overrides-test.d.ts [hunk 5]
```diff
@@ -358,9 +358,21 @@ export interface ExpectMatcherUtils {
 }
 
 export type ExpectMatcherState = {
+  /**
+   * Wehther this matcher was called with the negated .not modifier.
```
Comment (dgozman) [inline utils/generate_types/overrides-test.d.ts:362]:
```suggestion
   * Whether this matcher was called with the negated .not modifier.
```

File: packages/playwright/src/matchers/expect.ts [hunk 6]
```diff
@@ -129,7 +128,20 @@ function createExpect(info: ExpectMetaInfo) {
...
-          expectLibrary.extend(matchers);
+          const wrappedMatchers: any = {};
+          Object.entries(matchers).forEach(([name, matcher]) => {
+            wrappedMatchers[name] = function(...args: any[]) {
+              const { isNot, promise, utils } = this;
+              const newThis: ExpectMatcherState = {
+                isNot,
+                promise,
+                utils,
+                timeout: currentExpectTimeout({})
```
Comment (dgozman) [inline packages/playwright/src/matchers/expect.ts:139]: Might as well move `currentExpectTimeout` and `setCurrentExpectConfigureTimeout` into this file from the `globals.ts` and turn them local variables here.

**Review by dgozman — APPROVED** [review]

**Pull request merged**